pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -204,32 +269,251 @@ func (r *ReconcileNodeMaintenance) Reconcile(request reconcile.Request) (reconci
+			AcquireTime:          microTimeNow,
+			RenewTime:            microTimeNow,
+			LeaseTransitions:     makeInt32(0),
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+
+				nodeName := node.ObjectMeta.Name
+				nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+
+				if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+					return  nil, false, err
+				}
+				return lease, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func needUpdateOwnedLease(lease *coordv1beta1.Lease, currentTime time.Time) (bool,bool) {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		log.Info("empty renew time or duration in sec")
+		return true, true
+	}
+	dueTime := lease.Spec.RenewTime.Time
+	duration := *lease.Spec.LeaseDurationSeconds
+	dueTime = dueTime.Add(time.Duration(time.Duration(duration) * time.Second))
+
+	deadline := currentTime
+
+	// if lease expired right now  then both update the lease and the acquire time (second rvalue)
+	// if the acquire time has been previously nil
+	if dueTime.Before(deadline) {
+		return true, lease.Spec.AcquireTime == nil
+	}
+
+	deadline = deadline.Add(time.Duration(2 * drainerTimeoutInMinutes * time.Minute))
+
+	// about to expire, update the lease but no the acquire time (second rvalue)
+	return dueTime.Before(deadline), false
+}
+
+func isValidLease(lease *coordv1beta1.Lease, currentTime time.Time) bool {
+
+	if lease.Spec.RenewTime == nil || lease.Spec.LeaseDurationSeconds == nil {
+		return false
+	}
+
+	leaseDuration := time.Duration(int64(*lease.Spec.LeaseDurationSeconds) * int64(time.Second))
+	renewTime := (*lease.Spec.RenewTime).Time
+	dueTime := renewTime
+	dueTime = dueTime.Add(leaseDuration)
+
+	// valid lease if: due time not in the past and renew time not in the future
+	return !dueTime.Before(currentTime) && !renewTime.After(currentTime)
+}
+
+
+func updateLease(client  client.Client, node *corev1.Node, lease *coordv1beta1.Lease, currentTime time.Time, durationInSeconds int32) (*coordv1beta1.Lease, error, bool) {
+
+	holderIdentity := LeaseHolderIdentity
+	owner := &metav1.OwnerReference{
+			APIVersion: corev1.SchemeGroupVersion.WithKind("Node").Version,
+			Kind:       corev1.SchemeGroupVersion.WithKind("Node").Kind,
+			Name:       node.Name,
+			UID:        node.ObjectMeta.UID,
+	}
+
+	needUpdateLease := false
+	setAcquireAndLeaseTransitions := false
+	updateAlreadyOwnedLease := false
+
+	if  lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity == holderIdentity {
+		var needUpdate bool
+
+		needUpdate, setAcquireAndLeaseTransitions = needUpdateOwnedLease(lease,currentTime)
+		if  needUpdate {
+			needUpdateLease = true
+			updateAlreadyOwnedLease = true
+
+			log.Infof("renew lease owned by nmo setAcquireTime=%t", setAcquireAndLeaseTransitions)
+			lease.ObjectMeta.OwnerReferences = []metav1.OwnerReference{*owner}
```

Comment:
you can set these values in the `if needUpdateLease` block, it's duplicate here and in the `else` block

Reply:
done